rpc/handlers.go

```diff
@@ -208,87 +215,77 @@ type WsResponse struct {
+
+	evsw *events.EventSwitch
 }
 
 // new websocket connection wrapper
-func NewConnection(con *websocket.Conn) *Connection {
+func NewConnection(wsConn *websocket.Conn) *Connection {
 	return &Connection{
-		id:        con.RemoteAddr().String(),
-		wsCon:     con,
-		writeChan: make(chan WsResponse, WriteChanBuffer), // buffered. we keep track when its full
+		id:        wsConn.RemoteAddr().String(),
+		wsConn:    wsConn,
+		writeChan: make(chan WSResponse, WriteChanBufferSize), // buffered. we keep track when its full
 	}
 }
 
-// close the connection
-func (c *Connection) Close() {
-	c.wsCon.Close()
-	close(c.writeChan)
-	close(c.quitChan)
-}
+// start the connection and hand her the event switch
+func (con *Connection) Start(evsw *events.EventSwitch) {
+	if atomic.CompareAndSwapUint32(&con.started, 0, 1) {
+		con.evsw = evsw
 
-// main manager for all websocket connections
-// holds the event switch
-type WebsocketManager struct {
-	ew   *events.EventSwitch
-	cons map[string]*Connection
-}
-
-func NewWebsocketManager(ew *events.EventSwitch) *WebsocketManager {
-	return &WebsocketManager{
-		ew:   ew,
-		cons: make(map[string]*Connection),
+		// read subscriptions/unsubscriptions to events
+		go con.read()
+		// write responses
+		con.write()
 	}
 }
 
-func (w *WebsocketManager) eventsHandler(con *websocket.Conn) {
-	// register connection
-	c := NewConnection(con)
-	w.cons[con.RemoteAddr().String()] = c
-
-	// read subscriptions/unsubscriptions to events
-	go w.read(c)
-	// write responses
-	go w.write(c)
+// close the connection
+func (con *Connection) Stop() {
+	if atomic.CompareAndSwapUint32(&con.stopped, 0, 1) {
+		con.wsConn.Close()
+		close(con.writeChan)
+	}
 }
 
-const (
-	WsConnectionReaperSeconds = 5
-	MaxFailedSendsSeconds     = 10
-	WriteChanBuffer           = 10
-)
-
 // read from the socket and subscribe to or unsubscribe from events
-func (w *WebsocketManager) read(con *Connection) {
-	reaper := time.Tick(time.Second * WsConnectionReaperSeconds)
+func (con *Connection) read() {
+	reaper := time.Tick(time.Second * WSConnectionReaperSeconds)
 	for {
 		select {
 		case <-reaper:
-			if con.failedSends > MaxFailedSendsSeconds {
+			if con.failedSends > MaxFailedSends {
 				// sending has failed too many times.
 				// kill the connection
-				con.quitChan <- struct{}{}
+				con.Stop()
+				return
 			}
 		default:
 			var in []byte
-			if err := websocket.Message.Receive(con.wsCon, &in); err != nil {
+			_, in, err := con.wsConn.ReadMessage()
+			if err != nil {
 				// an error reading the connection,
-				// so kill the connection
-				con.quitChan <- struct{}{}
+				// kill the connection
+				con.Stop()
+				return
 			}
-			var req WsRequest
-			err := json.Unmarshal(in, &req)
+			var req WSRequest
+			err = json.Unmarshal(in, &req)
 			if err != nil {
 				errStr := fmt.Sprintf("Error unmarshaling data: %s", err.Error())
-				con.writeChan <- WsResponse{Error: errStr}
+				con.writeChan <- WSResponse{Error: errStr}
```

Comment:
need select{default:} here because writeChan couldhave been closed after wsConn is closed.
and more places prob need it too